<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="e0509_with_gripper">

  <!-- Arguments -->
  <xacro:arg name="color"        default="white"/>
  <xacro:arg name="namespace"    default="e0509_gripper"/>
  <xacro:arg name="use_gazebo"   default="false"/>
  <xacro:arg name="host"         default="127.0.0.1"/>
  <xacro:arg name="port"         default="12345"/>
  <xacro:arg name="rt_host"      default="192.168.137.50"/>
  <xacro:arg name="mode"         default="virtual"/>
  <xacro:arg name="model"        default="e0509"/>
  <xacro:arg name="update_rate"  default="100"/>

  <!-- Local properties -->
  <xacro:property name="gz"          value="$(arg use_gazebo)"/>
  <xacro:property name="host"        value="$(arg host)"/>
  <xacro:property name="rt_host"     value="$(arg rt_host)"/>
  <xacro:property name="port"        value="$(arg port)"/>
  <xacro:property name="mode"        value="$(arg mode)"/>
  <xacro:property name="model"       value="$(arg model)"/>
  <xacro:property name="update_rate" value="$(arg update_rate)"/>

  <!-- Include E0509 macro (includes gazebo plugin) -->
  <xacro:include filename="$(find dsr_description2)/xacro/macro.e0509.white.xacro"/>
  <xacro:e0509 namespace="$(arg namespace)"/>

  <!-- World link -->
  <link name="world"/>
  <joint name="world_fixed" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

  <!-- Include RH-P12-RN-A Gripper -->
  <xacro:include filename="$(find rh_p12_rn_a_description)/urdf/rh_p12_rn_a.xacro"/>
  <xacro:rh_p12_rn_a prefix="gripper_"/>

  <!-- Connect gripper to robot -->
  <joint name="gripper_attach_joint" type="fixed">
    <origin xyz="0 0.0 0.0" rpy="0 0 1.5708"/>
    <parent link="link_6"/>
    <child link="gripper_rh_p12_rn_base"/>
  </joint>

  <!-- Gazebo ros2_control -->
  <xacro:if value="${gz}">
    <!-- Arm ros2_control from dsr_description2 -->
    <xacro:include filename="$(find dsr_description2)/ros2_control/e0509.gz_ros2_control.xacro"/>
    <xacro:e0509_gz_ros2_control
      name="e0509_gripper_gz"
      namespace="$(arg namespace)"
      update_rate="${update_rate}"
    />
    
    <!-- Gripper ros2_control - ALL 4 joints -->
    <ros2_control name="gripper_gz" type="system">
      <hardware>
        <plugin>ign_ros2_control/IgnitionSystem</plugin>
      </hardware>
      <joint name="gripper_rh_r1">
        <command_interface name="position"/>
        <state_interface name="position"><param name="initial_value">0.0</param></state_interface>
        <state_interface name="velocity"/>
      </joint>
      <joint name="gripper_rh_r2">
        <command_interface name="position"/>
        <state_interface name="position"><param name="initial_value">0.0</param></state_interface>
        <state_interface name="velocity"/>
      </joint>
      <joint name="gripper_rh_l1">
        <command_interface name="position"/>
        <state_interface name="position"><param name="initial_value">0.0</param></state_interface>
        <state_interface name="velocity"/>
      </joint>
      <joint name="gripper_rh_l2">
        <command_interface name="position"/>
        <state_interface name="position"><param name="initial_value">0.0</param></state_interface>
        <state_interface name="velocity"/>
      </joint>
    </ros2_control>
  </xacro:if>

  <!-- Default ros2_control (Doosan hardware) -->
  <xacro:unless value="${gz}">
    <xacro:include filename="$(find dsr_description2)/ros2_control/e0509.ros2_control.xacro"/>
    <xacro:e0509_ros2_control
      name="${model}"
      host="${host}"
      rt_host="${rt_host}"
      port="${port}"
      mode="${mode}"
      model="${model}"
      update_rate="${update_rate}"
    />
  </xacro:unless>

</robot>
